docs/proposals/20221109-managed-external-etcd.md

```diff
@@ -0,0 +1,572 @@
+- This is not to be used for etcd-as-a-service. The etcd clusters created by an etcd provider within CAPI will only be used by the kube-apiserver of a target workload cluster.
+- Etcd cluster snapshots will need its own separate KEP. It should include details about how to take a snapshot, if `etcdctl snapshot save` command is used then how to collect the snapshots from the etcd machines and an integration with a storage provider to save those snapshots. This first iteration will not have a snapshot mechanism implemented. Users can still take snapshots by manually running the `etcdctl snapshot save` command and we can document those steps.
+
+## Proposal
+
+### User Stories
+
+- As an end user, I want to be able to create a Kubernetes cluster that uses external etcd topology using CAPI with a single step.
+- On creating the required CRs in CAPI along with the new etcd provider specific CRs, the external etcd provider should provision the etcd cluster first, and provide that as input to the workload cluster, without requiring any intervention from me as an end user.
+- As an end user, I should be able to use the etcd provider CRs to specify etcd cluster size during creation and specify/modify etcd version. CAPI controllers should modify and manage the etcd clusters accordingly.
+
+### Implementation Details/Notes/Constraints
+
+#### Data Model Changes
+
+The following new type/CRDs will be added for the etcdadm based provider
+```go
+// API Group: etcd.cluster.x-k8s.io OR etcdplane.cluster.x-k8s.io
+type EtcdadmCluster struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   EtcdadmClusterSpec   `json:"spec,omitempty"`
+    Status EtcdadmClusterStatus `json:"status,omitempty"`
+}
+
+// EtcdadmClusterSpec defines the desired state of EtcdadmCluster
+type EtcdadmClusterSpec struct {
+    // Replicas is the number of etcd members in the cluster. Defaults to 3 and immutable as really user shouldn't scale up/scale down without reason, since it will affect etcd quorum.
+    Replicas *int32 `json:"replicas,omitempty"`
+
+    // InfrastructureTemplate is a required reference to a custom resource
+    // offered by an infrastructure provider.
+    InfrastructureTemplate corev1.ObjectReference `json:"infrastructureTemplate"`
+
+    // +optional
+    EtcdadmConfigSpec etcdbp.EtcdadmConfigSpec `json:"etcdadmConfigSpec"`
+}
+
+// EtcdadmClusterStatus is the current state of the EtcdadmCluster
+type EtcdadmClusterStatus struct {
+    // Total number of non-terminated machines targeted by this etcd cluster
+    // (their labels match the selector).
+    // +optional
+    ReadyReplicas int32 `json:"replicas,omitempty"`
+
+    // +optional
+    InitMachineAddress string `json:"initMachineAddress"`
+
+    // +optional
+    Initialized bool `json:"initialized"`
+
+    // +optional
+    CreationComplete bool `json:"ready"`
```

Comment:
These should probably be conditions?

Reply:
Yeah, I would agree. Let me take a note and revisit this.